src/Service.Tests/Caching/DabCacheServiceIntegrationTests.cs

```diff
@@ -0,0 +1,345 @@
+using Moq;
+using ZiggyCreatures.Caching.Fusion;
+
+namespace Azure.DataApiBuilder.Service.Tests.Caching
+{
+    /// <summary>
+    /// Validates direct execution of DabCacheService.GetOrSetAsync(...)
+    /// Indirectly validates that the DabCacheService.CreateKey(...) method
+    /// creates consistent key values given the same input. The tests would fail
+    /// if key creation was invalid.
+    /// </summary>
+    [TestClass]
+    public class DabCacheServiceIntegrationTests
+    {
+        private const string ERROR_UNEXPECTED_INVOCATIONS = "Unexpected number of queryExecutor invocations.";
+        private const string ERROR_UNEXPECTED_RESULT = "Unexpected result returned by cache service.";
+        private const string ERROR_FAILED_ARG_PASSTHROUGH = "arg was not passed through to the executor as expected.";
+        private enum ExecutorReturnType { Json, Null, Exception };
+
+        /// <summary>
+        /// Validates that the first invocation of the cache service results in a cache miss because
+        /// the cache is expected to be empty.
+        /// After a cache miss, a call to the the factory method is expected. The factory method
+        /// is responsible for calling the database.
+        /// Validates that the cache service passes the expected metadata values through to the query executor
+        /// which indicates that the cache service does not modify request metadata.
+        /// </summary>
+        [TestMethod]
+        public async Task FirstCacheServiceInvocationCallsFactory()
+        {
+            // Arrange
+            using FusionCache cache = CreateFusionCache(sizeLimit: 1000, defaultEntryTtlSeconds: 1);
+            string expectedDatabaseResponseJson = @"{""key"": ""value""}";
+            Mock<IQueryExecutor> mockQueryExecutor = CreateMockQueryExecutor(rawJsonResponse: expectedDatabaseResponseJson, ExecutorReturnType.Json);
+
+            Dictionary<string, DbConnectionParam> parameters = new()
+            {
+                {"param1", new DbConnectionParam(value: "param1Value") }
+            };
+
+            DatabaseQueryMetadata queryMetadata = new(queryText: "select * from MyTable", dataSource: "dataSource1", queryParameters: parameters);
+            DabCacheService dabCache = CreateDabCacheService(cache);
+
+            // Act
+            int cacheEntryTtl = 1;
+            JsonElement? result = await dabCache.GetOrSetAsync<JsonElement?>(queryExecutor: mockQueryExecutor.Object, queryMetadata: queryMetadata, cacheEntryTtl: cacheEntryTtl);
+
+            // Assert
+            Assert.AreEqual(expected: true, actual: mockQueryExecutor.Invocations.Count is 1, message: ERROR_UNEXPECTED_INVOCATIONS);
+
+            IReadOnlyList<object> actualExecuteQueryAsyncArguments = mockQueryExecutor.Invocations[0].Arguments;
+            Assert.AreEqual(expected: queryMetadata.QueryText, actual: actualExecuteQueryAsyncArguments[0], message: "QueryText " + ERROR_FAILED_ARG_PASSTHROUGH);
```

Comment:
Would these indices change if their order is modified in the executor? If yes, Can we somehow enforce this modification to indices if the order of arguments changes?

Reply:
They would change if the method signature of the `queryExecutor` changed the order. In that case the test would break because Moq does not support setting up a mock with named parameters (which may help mitigate the issue you highlight). 

Moq setup looks like the following with no named params supported (compiler error):
```csharp
                    mockQueryExecutor.Setup(x => x.ExecuteQueryAsync(
                        It.IsAny<string>(),
                        It.IsAny<IDictionary<string, DbConnectionParam>>(),
                        It.IsAny<Func<DbDataReader?, List<string>?, Task<JsonElement?>>>(),
                        httpContext,
                        args,
                        It.IsAny<string>()).Result)
                        .Returns((JsonElement?)null);
```

Additionally, the MOQ object's property `mockQueryExecutor.Invocations[0].Arguments` is a readonly list of objects. There is no key representing the parameter/argument name. I would only be able to check what object type each list entry is, but that falls short when you have multiple params of the same type (e.g. string).